### PR TITLE
Layertree refactor

### PR DESCRIPTION
**Pull request by ger-benjamin**
Fix: #1752 

Main changes:
- There is a state in each node of the layertree.
- Set state of a layertree and ask to sync will sync the layers based on the tree.
- The creation of layers was totally refactored (a layer is now created / managed when the corresponding layertree is created).
- A syncLayertreeMap service create the layers and sync the tree.
- Externs are ready to reflect now the server response.
- Layertree is now well tested. 
- We easily can get a ngeo.Layeree based on it's node id (but it retrieve the first found !) (then active or deactive it's state, like that: https://github.com/camptocamp/ngeo/blob/layertree_refactor/contribs/gmf/src/services/treemanager.js#L357-L375)

# TODO for this PR

## Help needed
**Features that must be fixed before merge or in other issues**:
 - Permalink (strange behaviors with multiple groups, try with the edit theme)
 - ~~Edition (problem cause now, layers are created/updated one by one)~~ fixed :-)

# TODO After this PR

## Issues opened
 - #1885

## To check / improve

We probably can move some things in the syncLayertreeMap service. Like:
 - Dimension management (currently in gmf.Layertree)
 - Opacity (currently in gmf.Layertree)

Perhaps, things can be simplified in these files (files where there was gmfThemesNode object):
 - contribs/gmf/service/querymanager.js
 - contribs/gmf/service/themeservice.js
 - contribs/gmf/service/treemanager.js
 - contribs/gmf/service/permalink.js
 - contribs/gmf/directive/themeselector.js
 - contribs/gmf/directive/layertree.js
 - contribs/gmf/directive/editfeatueselector.js
 - contribs/gmf/directive/editfeatue.js

Others files to check: 
 - src/service/layerhelper.js
 - src/directive/layertree.js

And A lot of `gmfWmsUrl` contantes exist, it's not longer used like in layertree. Must be removed

### CODE REVIEW

**Comment by ger-benjamin** [comment]
@adube If you have time for that, can you take this branch and try to debug the permalink ?

One obvious problem come when you load the displayquerywindow example:
At this line: https://github.com/camptocamp/ngeo/blob/master/contribs/gmf/src/services/permalink.js#L783 we get a `true` value. That's obviously unexpected and the example is broken...

The main change that can affect the code is that the gmf layertree creates or update layer on by one (via the new the gmf.SyncLayertreeMap service). (that also create an awkward error to collect the editable layers).

Thanks. And sorry, I need to go home now...

PS: (karma tests can't pass for now. But example must)

**Comment by pgiraud** [comment]
Hi @fredj & @ger-benjamin I just rebased the branch to fix the conflicts with changes in master.

File: contribs/gmf/src/directives/editfeature.js [hunk 1]
```diff
@@ -183,6 +174,24 @@ gmf.EditfeatureController = function($element, $scope, $timeout, $q,
...
 
   /**
+   * @type {GmfThemesLeaf}
+   * @private
+   */
+  this.editableNode_ = /** @type {GmfThemesLeaf} */ (
+    this.editableTreeCtrl.node);
+
+
+  var layer = this.editableTreeCtrl.parent.layer;
```
Comment (pgiraud) [inline contribs/gmf/src/directives/editfeature.js]: What if level is > 1?

Comment (adube) [inline contribs/gmf/src/directives/editfeature.js]: You're right.  I need to fix this.

**Comment by gberaudo** [comment]
We had a discussion between @pgiraud, @sbrunner and @fredj.
Here are the points we want to address before merging:

- Move SyncLayerTreeMap service to the TreeManager service;
- Emit events on layer tree state changes and listen to them in consumer code (permalink, sync layertree) instead of having watchers or direct call (sync function);
- Move low level code about first layer group and mixed state from sync code to event emiting code, the OL3 layer is stored at creation time in the appropriate layer controller;
- Rename layertreeCtrl node to `initialConfig` (read only);
- Move layertreeCtrl.node['editing'] to layerCtrl.editing;
- For edit feature, the treeCtrlReferences must not be used; Instead, we should watch for changes of the first layer group list;
- The treeCtrlReferences function is renamed and only the first layer group controllers are kept in the array.

File: contribs/gmf/examples/editfeatureselector.html [hunk 2]
```diff
@@ -94,6 +94,12 @@
         class="panel panel-default panel-body">
     </gmf-authentication>
 
+    <gmf-layertree
```
Comment (gberaudo) [inline contribs/gmf/examples/editfeatureselector.html]: Keep this invisible layer tree?

File: contribs/gmf/src/directives/editfeature.js [hunk 3]
```diff
@@ -183,6 +175,24 @@ gmf.EditfeatureController = function($element, $scope, $timeout, $q,
...
 
   /**
+   * @type {GmfThemesLeaf}
+   * @private
+   */
+  this.editableNode_ = /** @type {GmfThemesLeaf} */ (
+    this.editableTreeCtrl.node);
+
+
+  var layer = gmf.SyncLayertreeMap.getLayer(this.editableTreeCtrl);
```
Comment (gberaudo) [inline contribs/gmf/src/directives/editfeature.js]: Get the layer with `this.editableTreeCtrl.layer`.

File: contribs/gmf/src/directives/editfeatureselector.js [hunk 4]
```diff
@@ -112,81 +96,80 @@ gmf.EditfeatureselectorController = function($scope, gmfThemes,
...
-   * @export
+   * @type {gmf.TreeManager}
+   * @private
    */
-  this.editableNodes = [];
+  this.gmfTreeManager_ = gmfTreeManager;
+
+  this.treeCtrlReferencesWatcherUnregister_ = $scope.$watchCollection(
+    function() {
+      return this.gmfTreeManager_.getTreeCtrlReferences();
```
Comment (gberaudo) [inline contribs/gmf/src/directives/editfeatureselector.js]: Listen `this.gmfTreeManager_.getFirstLevelLayerTreeCtrls()`

File: contribs/gmf/src/directives/layertree.js [hunk 5]
```diff
@@ -272,253 +285,38 @@ gmf.LayertreeController.prototype.updateLayerDimensions_ = function(layer, node)
...
-    timeValues = this.gmfWMSTime_.getOptions(wmsTime)['values'];
-    timeParam = this.gmfWMSTime_.formatWMSTimeParam(wmsTime, {
-      start : timeValues[0] || timeValues,
-      end : timeValues[1]
-    });
+  var opt_position;
+  // Precise the index to add first level groups.
+  if (treeCtrl.parent.isRoot) {
+    opt_position = this.gmfTreeManager_.tree.children.length -
+        this.gmfTreeManager_.layersToAddAtOnce | 0;
```
Comment (gberaudo) [inline contribs/gmf/src/directives/layertree.js]: Suspicious binary or

File: src/directives/layertree.js [hunk 6]
```diff
@@ -241,14 +264,93 @@ ngeo.LayertreeController = function($scope, $element, $attrs, ngeoDecorateLayer,
...
+
+/**
+ * Set the state of this treeCtrl. Update its children with its value and then
+ * ask its parent to refresh its state.
+ * @param {string} state 'on' or 'off'.
+ * @param {boolean=} opt_avoidRefreshParents True to avoid refreshing its
+ * parent.
+ * @export
+ */
+ngeo.LayertreeController.prototype.setState = function(state,
```
Comment (gberaudo) [inline src/directives/layertree.js]: Use a setStateInternal to only change the state.
At the end setState, emit a change event (for the whole branch).

File: contribs/gmf/examples/editfeatureselector.html (hunk 2 above)
Comment (adube) [inline contribs/gmf/examples/editfeatureselector.html]: Without a gmf-layertree, layers and nodes are not created.

File: contribs/gmf/src/directives/layertree.js [hunk 7]
```diff
@@ -562,132 +363,9 @@ gmf.LayertreeController.prototype.getResolutionStyle = function(node) {
...
-    case gmf.Themes.NodeType.WMS:
-      return node.layers;
-    case gmf.Themes.NodeType.WMTS:
-      return node.layer;
-    default:
-      throw new Error('Node wrong type to get layer(s) name(s): ' + type);
-  }
+  treeCtrl.setState(treeCtrl.getState() === 'on' ? 'off' : 'on');
+  var firstLevelTreeCtrl = ngeo.LayertreeController.getFirstParentTree(treeCtrl);
+  this.gmfSyncLayertreeMap_.sync(this.map, firstLevelTreeCtrl);
```
Comment (gberaudo) [inline contribs/gmf/src/directives/layertree.js]: Remove this call, it will be done in setState.

File: contribs/gmf/src/services/treemanager.js [hunk 8]
```diff
@@ -76,15 +85,21 @@ gmf.TreeManager = function($timeout, gettextCatalog, ngeoLayerHelper,
...
+  this.tree = /** @type {GmfThemesTheme} */ ({
     children: [],
     name: ''
   });
 
   /**
+   * @type {Array.<ngeo.LayertreeController>}
+   * @private
+   */
+  this.treeCtrlReferences_ = [];
```
Comment (gberaudo) [inline contribs/gmf/src/services/treemanager.js]: Only keep the first level layers

File: contribs/gmf/src/services/treemanager.js [hunk 9]
```diff
@@ -332,14 +347,34 @@ gmf.TreeManager.prototype.addGroupByName = function(groupName, opt_add,
...
     if (group) {
       var groupAdded = this.addGroups([group], opt_add, opt_silent);
       if (opt_map) {
+        var map = opt_map;
         this.$timeout_(function() {
-          this.setLayerVisible_(layerName, group, groupAdded, /** @type {ol.Map}*/ (opt_map));
+          var treeCtrl = this.getTreeCtrlByNodeId(group.id);
+          var treeCtrls = [];
+          var treeCtrlToActive;
+          ngeo.LayertreeController.getFlatTree(treeCtrl, treeCtrls);
```
Comment (gberaudo) [inline contribs/gmf/src/services/treemanager.js]: Don't use flatTree

File: contribs/gmf/src/services/treemanager.js [hunk 10]
```diff
@@ -504,4 +475,57 @@ gmf.TreeManager.prototype.notifyCantAddGroups_ = function(groups) {
...
+  return this.treeCtrlReferences_;
+};
+
+
+/**
+ * Add a treeCtrl to kept in reference.
+ * @param {ngeo.LayertreeController} treeCtrl ngeoLayer tree controller.
+ * @public
+ */
+gmf.TreeManager.prototype.addTreeCtrlReference = function(treeCtrl) {
```
Comment (gberaudo) [inline contribs/gmf/src/services/treemanager.js]: Remove and directly assign the level 1 array.

File: contribs/gmf/src/services/treemanager.js [hunk 11]
```diff
@@ -504,4 +475,57 @@ gmf.TreeManager.prototype.notifyCantAddGroups_ = function(groups) {
...
+  }
+};
+
+
+/**
+ * Remove a treeCtrl kept as reference.
+ * @param {ngeo.LayertreeController} treeCtrl ngeoLayer tree controller.
+ * @public
+ */
+gmf.TreeManager.prototype.removeTreeCtrlReference = function(treeCtrl) {
```
Comment (gberaudo) [inline contribs/gmf/src/services/treemanager.js]: Remove, the whole array will be replaced.

File: contribs/gmf/src/services/treemanager.js [hunk 12]
```diff
@@ -504,4 +475,57 @@ gmf.TreeManager.prototype.notifyCantAddGroups_ = function(groups) {
...
+/**
+ * Get a treeCtrl based on it's node id. The treeCtrl must have been added to
+ * this service to be found.
+ * @param {number} id the id of a GMFThemesGroup or a GMFThemesLeaf.
+ * @return {ngeo.LayertreeController} treeCtrl Todo.
+ * @public
+ */
+gmf.TreeManager.prototype.getTreeCtrlByNodeId = function(id) {
+  var correspondingTreeCtrl = null;
+  this.treeCtrlReferences_.some(function(treeCtrl) {
```
Comment (gberaudo) [inline contribs/gmf/src/services/treemanager.js]: Use a visitor.

File: src/directives/layertree.js [hunk 13]
```diff
@@ -241,14 +264,93 @@ ngeo.LayertreeController = function($scope, $element, $attrs, ngeoDecorateLayer,
...
+};
+
+
+/**
+ * Refresh the state of this treeCtrl based on it's children value. The call its
+ * parent to do the same.
+ * @public
+ */
+ngeo.LayertreeController.prototype.refreshState = function() {
+  var newState = this.getCalculateState();
```
Comment (gberaudo) [inline src/directives/layertree.js]: This looks like it could be optimized.

File: src/directives/layertree.js [hunk 14]
```diff
@@ -261,4 +363,43 @@ ngeo.LayertreeController.prototype.getSetActive = function(val) {
...
+/**
+ * Fill the given array with all layertree objects of any level from the
+ * children of the given layertree. By default, include onyl all leaves.
+ * @param {ngeo.LayertreeController} treeCtrl ngeo layertree controller.
+ * @param {Array.<ngeo.LayertreeController>} treeCtrls array that will contains
+ *     the ngeo layertree controller.
+ * @param {boolean=} opt_includeGroups If true, include leaves and groups.
+ * @public
+ */
+ngeo.LayertreeController.getFlatTree = function(treeCtrl, treeCtrls,
```
Comment (gberaudo) [inline src/directives/layertree.js]: Must be removed.

File: src/services/layerHelper.js [hunk 15]
```diff
@@ -219,6 +220,38 @@ ngeo.LayerHelper.prototype.getFlatLayers_ = function(layer, array) {
 
 
 /**
+ * Get an array of all group and layer in a group. The group can contain
+ * multiple levels of others groups. It will include itself.
+ * @param {ol.layer.Group} group The base group of layers
+ * @return {Array.<ol.layer.Group|ol.layer.Base>} .
+ * @export
+ */
+ngeo.LayerHelper.prototype.getFlatMapElements = function(group) {
```
Comment (gberaudo) [inline src/services/layerHelper.js]: Is it used?

File: src/services/layerHelper.js [hunk 16]
```diff
@@ -219,6 +220,38 @@ ngeo.LayerHelper.prototype.getFlatLayers_ = function(layer, array) {
...
+
+/**
+ * Get an array of all layers and group in a group. The group can contain
+ * multiple levels of others groups.
+ * @param {ol.layer.Group|ol.layer.Base} element A group or a layer.
+ * @param {Array.<ol.layer.Group|ol.layer.Base>} array An array to add elements.
+ * @return {Array.<ol.layer.Group|ol.layer.Base>} Elements.
+ * @private
+ */
+ngeo.LayerHelper.prototype.getFlatMapElements_ = function(element, array) {
```
Comment (gberaudo) [inline src/services/layerHelper.js]: Is it used?

File: contribs/gmf/src/directives/layertree.js [hunk 17]
```diff
@@ -190,6 +190,13 @@ gmf.LayertreeController = function($http, $sce, $scope, ngeoCreatePopup,
   this.groupNodeStates_ = {};
 
   /**
+   * Flag that is turned on when the leaf is currently being edited.
+   * @type {boolean}
+   * @export
+   */
+  this.editing = false;
```
Comment (fredj) [inline contribs/gmf/src/directives/layertree.js]: FIXME: the property must be moved to `ngeo.LayertreeController`

Comment (fredj) [inline contribs/gmf/src/directives/layertree.js]: fixed with 8a5c48f

**Comment by fredj** [comment]
permalink issue: cadastre group is loaded

File: src/services/layerHelper.js (hunk 15 above)
Comment (fredj) [inline src/services/layerHelper.js]: not used

File: contribs/gmf/src/services/treemanager.js (hunk 11 above)
Comment (fredj) [inline contribs/gmf/src/services/treemanager.js]: done

**Pull request merged**